### PR TITLE
Memory game initial card state generation

### PR DESCRIPTION
**Pull request by tmansuroglu**
These are parts of bigger code. To prevent test error, parts related to other PRs are commented out. Screenshot not added because it is not meaningful without other components

#22

### CODE REVIEW

File: src/components/memoryCardBackSideImages.jsx
Comment (louisrli) [inline src/components/memoryCardBackSideImages.jsx]: CARD_BACKSIDE_IMAGES

File: src/components/memoryCardBackSides.jsx
Comment (louisrli) [inline src/components/memoryCardBackSides.jsx]: generally we want to avoid having "parallel lists" like this. though the comment is good, still that type of code (you will find with experience, i guess i encourage you to try it in another project to feel the pain of it) is hard to maintain and prone to error

what i would do:
* install the "invariant" package in npm
* `invariant(cardData.length === cardBacksideImages.length)`
* `const cardDataAndBacksides = []`. push [cardData[0], cardBacksideImages[0]] to the list...so it's a list of lists.
* you can now safely loop through cardDataAndBacksides such that its paired correctly; bc of your invariant you are sure that things are paired together correctly

File: src/components/memoryCardBackSides.jsx
Comment (louisrli) [inline src/components/memoryCardBackSides.jsx]: avoid using index as key

File: src/containers/MemoryGame/index.jsx
Comment (louisrli) [inline src/containers/MemoryGame/index.jsx]: no commented code in PRs as a general rule

File: src/containers/MemoryGame/index.jsx
Comment (louisrli) [inline src/containers/MemoryGame/index.jsx]: as a general rule of thumb as a programmer you should always think this way: "if i have 5 things, how hard would it be for me to make it 10000 things?" the answer in this code is, very hard, because you need to add 9995 more states.

this code needs to be written such that it works for an arbitrary number N cards (and it is possible). so when you rewrite this code, think very very carefully about this rule of thumb i've given (and most code you write in your life from now on, there are exceptions where sometimes its faster to write things for the case of 2 or 3 but never for the case of 10-15 like this).

to give you a start, what you probably want to do is:
```
const cardFlippedState = React.useState([new Array(cardBackImages.length).fill(false)]);

const flipCard = (index: number) => {
   // mutate the state here
}

```
you can also consider making your state an object mapping image names to true/false rather than using indexes (cleaner and easier to debug IMO)

how i personally would do it, in the most extensible / flexible way to write code possible:
* cardSpecs.jsx. contains data about cards, contains an object `const CARD_SPECS` that has keys "string" and maps to "objects". objects includes backImage, but its extensible to include more data later if i need to add more data about cards (say "hints" or something)
* cardFlippedState would be an object that maps strings to booleans. and i would make it by looping through Object.keys(CARD_SPECS) and making it false by default
* flipCard would take a cardKey name, not an index (working with indexes is very unreadable in general)
* CARD_SPECS again is the source of truth and can be used to render all things on the page

but i want to emphasize this: never after today should you write code that has this many states that do the same thing. it's not one of those things where it's like "oh sometimes there's exceptions."

File: src/containers/MemoryGame/index.jsx
Comment (tmansuroglu) [reply src/containers/MemoryGame/index.jsx]: Line 92 will be used when it is merged with other PRs. Should I still delete it?

File: src/components/memoryCardBackSideImages.jsx
Comment (tmansuroglu) [reply src/components/memoryCardBackSideImages.jsx]: PTAL

File: src/components/memoryCardBackSides.jsx
Comment (tmansuroglu) [reply src/components/memoryCardBackSides.jsx]: `key `is already `cardData.description`.

If you meant `id={index}`, the reason behind it is that matching cards should have same id. Since there should be 2 of each card, after creating array with this function, i duplicate same array and put old and new array together in one array. I could do the same by creating ids by hand. But this way felt more logical. Should I still remove it?

File: src/components/memoryCardBackSides.jsx
Comment (tmansuroglu) [reply src/components/memoryCardBackSides.jsx]: PTAL

File: src/components/memoryCardBackSides.jsx
Comment (majdajroudi) [inline src/components/memoryCardBackSides.jsx]: I think the convention here is CardBackSideImages or CardBacksideImages, I'm not sure but the first letter for a component should be in caps

File: src/components/memoryCardBackSides.jsx
Comment (louisrli) [reply src/components/memoryCardBackSides.jsx]: sorry i misread but you should also avoid using index as id. first, the id would be just "0", "1", "2", which is unreadable; IDs are global. second, ID shouldn't be used anyway except in very very rare situations. it also doesn't make sense that matching cards should have the same ID because IDs are globally unique, you can only have one div with that id

Comment (tmansuroglu) [reply src/components/memoryCardBackSides.jsx]: PTAL

File: src/containers/MemoryGame/index.jsx
Comment (tmansuroglu) [reply src/containers/MemoryGame/index.jsx]: PTAL

File: src/containers/MemoryGame/index.jsx
Comment (tmansuroglu) [reply src/containers/MemoryGame/index.jsx]: Now code is reusable.
` const initialStates = {};
cardsArr.forEach((card) => {
    initialStates[card.key] = false;
  });
const [cardStates, setCardStates] = useState(initialStates);`

State gets manipulated with `setCardStates(prevState=>{...prevState,key:state})`

Also moved code to a more appropriate file. In addition I discovered the reason behind my first issue. Before using 20 states, I was trying to write something that is reusable(1 state to control all cards). After trying for 4.5 hours I couldn't solve problem of inaccurate flips and twitches and decided to create individual states. Today I was able to recreate same issue by passing state to children. When I moved `cardStates` where it is being used issue resolved. I don't know the exact reason behind it. But i plan to find out soon.

PTAL

File: src/components/memoryCardBackSideImages.jsx
Comment (louisrli) [inline src/components/memoryCardBackSideImages.jsx]: this file should be under src/components/memorygame/BackSideImages or something like this, its polluting the components space right now

also if you are going to hold info for the cards other than the backside images i would very much highly recommend doing the object approach that i said earlier (aka if you have a frontside or other facts associated with a card, i would not accept code that keeps them separately and correlates them using the array indexes)

the reason is that its extremely bug prone and you're relying on a human not to swap the order accidentally in one of the arrays. who is to say youre not going to make alternateEnergy 3rd instead of 4th accidentally in the front, but 4th instead of 3rd in the array for the back?

i see its in alphabetical order right now but in the end youre relying on a human to keep your things synchronized. again, think about if you had 10,000 cards

File: src/components/shuffle.jsx
Comment (louisrli) [inline src/components/shuffle.jsx]: it should be clarified that this is not an in-place shuffle in this comment, which would be the first thing a reader wants to know when they see a fn called shuffle. not whether or not
 the fn called shuffle "Shuffles cards" (comments should be used to clarify non-obvious things or mark headers in longer files)

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: if you are returning jsx here, then this whole thing should be a component, not a function

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: this is called Arr but it's a function

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: needs comments why there are clones. imagine you are me. why are there clones?

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: are not matching length

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: shuffledCardDataBacksidesClones

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: already something is wrong here when one thing returns a boolean but the other returns an empty string

Comment (louisrli) [reply src/components/memoryCardBackSidesArr.jsx]: two different types

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: so if you just make one file called cardspecs that has all the data of your cards, you can basically delete 80% of this code and just loop through one object. is there a particular reason card backs need to be separated from card fronts? in a real database, you would have them all stored together in a row/table (or at least join them together to get a single logical object)

File: src/components/memoryCardBackSidesArr.jsx
Comment (louisrli) [inline src/components/memoryCardBackSidesArr.jsx]: i think you need to include in the PR what the game is, in addition to writing comments, again imagine you are your teammate or someone who joins this team trying to understand what's going on with the cloning thing

File: src/components/memoryCardBackSidesArr.jsx
Comment (tmansuroglu) [reply src/components/memoryCardBackSidesArr.jsx]: To make flipping work, you need define 2 children(1 front ,1 backside) under `<ReactCardFlip></ReactCardFlip>`. Also, I didn't store them inside 1 object at first because im getting the data from i18n. Front image is same for all and backside images are repeated per language. But after getting data from i18n, I can create and object to hold everything.

Comment (louisrli) [reply src/components/memoryCardBackSidesArr.jsx]: so there are a few options here that would be cleaner than what's currently here
* you store them inside 1 object but you stored the i18n translation keys instead of directly storing the data
* you create another virtual layer, processing after translation, to create objects that you can access (a class would be suitable here)

compare these two, and imagine you have to add more data eventually

```
    const { description } = dataAndImage[0];
    const image = dataAndImage[1];
    const anotherThing = dataAndImage[2];
```

vs code that looks like this (your goal)
```
console.log(cardData.image)
console.log(cardData.description)
console.log(cardData.title)
```

basically, if you can't define it upfront, you want to have a preprocessing step so that you (and your teammate, and your future self, who have to read / understand the code) can work with a cleanly defined cardData object

Comment (louisrli) [reply src/components/memoryCardBackSidesArr.jsx]: i'm not understanding the cardflip thing and cloning. unless i'm misunderstanding the game. and even if you need to clone it, you shouldn't clone it at the time of preprocessing, you should clone it at the time the clone is needed (creating clones of data and passing them around gets extremely confusing).

```
const [isFlipped, setIsFlipped] = React.useState(false)

<div>
{!isFlipped && <CardFront data={cardData.front}/>}
{isFlipped && <CardBack data={cardData.back}/>}
</div>
```

Comment (louisrli) [reply src/components/memoryCardBackSidesArr.jsx]: with cloning at usage time:
```
const [isFlipped, setIsFlipped] = React.useState(false)
const cardDataClone = // clone card data
<div>
{!isFlipped && <CardFront data={cardData.front}/>}
{isFlipped && <CardBack data={cardDataClone.back}/>}
</div>
```
or something like this, if for some reason the data needs to be cloned

Comment (tmansuroglu) [reply src/components/memoryCardBackSidesArr.jsx]: Im cloning the cards because the purpose of game is matching cards with same backside image. Similar example is [this](https://www.brightfocus.org/alzheimers/memory-games/match-pictures). But instead of making square disappear after clicking on different sqaure, I made them disappear after certain amount of time. So you click on card it flips and shows what is behind and after some time it flicks back to show front side.

Comment (tmansuroglu) [reply src/components/memoryCardBackSidesArr.jsx]: First I will create an object that holds everything. And I will separate object creating step to a different file to make it obvious.

Comment (tmansuroglu) [reply src/components/memoryCardBackSidesArr.jsx]: Also clones actually are not identical. They have different keys. But I will try to implement what you suggested. I can implement it by lifting up `selectedCards`

File: src/components/memorygame/cardInitialStateGenerator.jsx
Comment (louisrli) [inline src/components/memorygame/cardInitialStateGenerator.jsx:7]: nicely done. if we were using typescript, we wouldn't need to do this because the types would be clearly declared, but in javascript we have to do things like this. this is also not ideal because these comments don't age well in case you change the structure but there's not much that can be done about that

File: src/components/memorygame/cardInitialStateGenerator.jsx
Comment (louisrli) [inline src/components/memorygame/cardInitialStateGenerator.jsx]: fn parameters should never be ALL_CAPS, numRequiredMatches (it doesn't know whether or not its taking a constant, it doesnt care, teh function just knows it accepts a number)

File: src/components/memorygame/cardInitialStateGenerator.jsx
Comment (louisrli) [inline src/components/memorygame/cardInitialStateGenerator.jsx]: can delete this comment, it just needs to be in the declaration of shuffle (if i see this without the comment i assume its nondestructive because you return it)

File: src/components/memorygame/shuffle.jsx
Comment (louisrli) [inline src/components/memorygame/shuffle.jsx:4]: nondestructive shuffle goes here; think about whether it makes more sense to comment "non-destructive shuffle" above every call to this function or in the place the function is declared (if i used htis 100x in my codebase and wrote it above every time, that would seem a bit tedious)

File: src/containers/MemoryGame/index.jsx
Comment (louisrli) [inline src/containers/MemoryGame/index.jsx]: i don't think you should be having to declare the Navbar on every page...seems wrong. it should be declared in a top-level component, and using routing the thing you want to show should show up.

File: src/components/memorygame/cardInitialStateGenerator.jsx
Comment (louisrli) [inline src/components/memorygame/cardInitialStateGenerator.jsx]: ok...it took me a while to get this code, its doing something tricky which is that, as it goes through the forEach and builds more cards, then numOfSameCardKeys gets incremented, almost implicitly (if there are 6 acid rains and i'm through 3 of them, then this finds the previous 2).

i would call this variable numSameCardKeysSoFar and add a comment that this does the same thing as incrementing by doing this search

**Review by louisrli — APPROVED** [review]
generally good, minor comments pls fix before merging

File: src/components/memorygame/cardInitialStateGenerator.jsx
Comment (tmansuroglu) [reply src/components/memorygame/cardInitialStateGenerator.jsx]: Added

File: src/containers/MemoryGame/index.jsx
Comment (tmansuroglu) [reply src/containers/MemoryGame/index.jsx]: fixed

File: src/components/memorygame/shuffle.jsx
Comment (tmansuroglu) [reply src/components/memorygame/shuffle.jsx:4]: fixed

File: src/components/memorygame/cardInitialStateGenerator.jsx
Comment (tmansuroglu) [reply src/components/memorygame/cardInitialStateGenerator.jsx]: fixed

File: src/components/memorygame/cardInitialStateGenerator.jsx
Comment (tmansuroglu) [reply src/components/memorygame/cardInitialStateGenerator.jsx]: fixed

**Comment by tmansuroglu** [comment]
@majdajroudi @ssinikoglu @ibrahimaljarad  guys i need review for approval.

File: src/components/memoryCardBackSides.jsx
Comment (majdajroudi) [reply src/components/memoryCardBackSides.jsx]: This is an old comment that I wrote and never submitted, please ignore it

**Review by majdajroudi — APPROVED** [review]
very well documented. I will use your shuffle function in my PR

**Comment by tmansuroglu** [comment]
@Ammar-64 can you merge?

**Pull request merged**
